File: pyaff4/stream_test.py
import data_store
import os
import lexicon
import rdfvalue
import unittest
import plugins


class StreamTest(unittest.TestCase):
    def streamTest(self, stream):
        self.assertEquals(0, stream.Tell())
        self.assertEquals(0, stream.Size())

        stream.Write("hello world")
        self.assertEquals(11, stream.Tell())

        stream.Seek(0, 0)
        self.assertEquals(0, stream.Tell())

        self.assertEquals("hello world",
                          stream.Read(1000))

        self.assertEquals(11, stream.Tell())

        stream.Seek(-5, 2)
        self.assertEquals(6, stream.Tell())

        self.assertEquals("world",
                          stream.Read(1000))

        stream.Seek(-5, 2)
        self.assertEquals(6, stream.Tell())

        stream.Write("Cruel world")
        stream.Seek(0, 0)
        self.assertEquals(0, stream.Tell())
        self.assertEquals("hello Cruel world",
                          stream.Read(1000))

        self.assertEquals(17, stream.Tell())

        stream.Seek(0, 0)

        self.assertEquals("he",
                          stream.Read(2))

        stream.Write("I have %d arms and %#x legs." % (2, 1025))
        self.assertEquals(31, stream.Tell())

        stream.Seek(0, 0)
        self.assertEquals("heI have 2 arms and 0x401 legs.",
                          stream.Read(1000))

    def testFileBackedStream(self):
        filename = rdfvalue.URN("/tmp/test_filename.bin")
        resolver = data_store.MemoryDataStore()
        try:
            resolver.Set(filename, lexicon.AFF4_STREAM_WRITE_MODE,
                         rdfvalue.XSDString("truncate"))

            file_stream = resolver.AFF4FactoryOpen(filename)
            self.streamTest(file_stream)
        finally:
            os.unlink(filename.Parse().path)


if __name__ == '__main__':
    unittest.main()

File: tests/stream_test.cc
#include <gtest/gtest.h>
#include "aff4/libaff4.h"
#include <unistd.h>

namespace aff4 {


class StreamTest : public ::testing::Test {
 protected:
  void test_Stream(AFF4Stream &stream) {
    EXPECT_EQ(0, stream.Tell());
    EXPECT_EQ(0, stream.Size());

    stream.Write("hello world");
    EXPECT_EQ(11, stream.Tell());

    stream.Seek(0, 0);
    EXPECT_EQ(0, stream.Tell());
    EXPECT_STREQ("hello",
                 stream.Read(5).c_str());
    EXPECT_EQ(5, stream.Tell());

    stream.Seek(0, 0);
    EXPECT_STREQ("hello world",
                 stream.Read(1000).c_str());

    EXPECT_EQ(11, stream.Tell());

    stream.Seek(-5, 2);
    EXPECT_EQ(6, stream.Tell());

    EXPECT_STREQ("world",
                 stream.Read(1000).c_str());

    stream.Seek(-5, 2);
    EXPECT_EQ(6, stream.Tell());

    stream.Write("Cruel world");
    stream.Seek(0, 0);
    EXPECT_EQ(0, stream.Tell());
    EXPECT_STREQ("hello Cruel world",
                 stream.Read(1000).c_str());

    EXPECT_EQ(17, stream.Tell());

    stream.Seek(0, 0);

    EXPECT_STREQ("he",
                 stream.Read(2).c_str());

    stream.sprintf("I have %d arms and %#x legs.", 2, 1025);
    EXPECT_EQ(31, stream.Tell());

    stream.Seek(0, 0);
    EXPECT_STREQ("heI have 2 arms and 0x401 legs.",
                 stream.Read(1000).c_str());

  };

};

TEST_F(StreamTest, StringIOTest) {
        std::unique_ptr<AFF4Stream> stream = StringIO::NewStringIO();

  test_Stream(*stream);
}

class FileBackedStreamTest: public StreamTest {
 protected:
        std::string filename = "/tmp/test_filename.bin";

  // Remove the file on teardown.
  virtual void TearDown() {
    unlink(filename.c_str());
  };
};

TEST_F(FileBackedStreamTest, FileBackedObjectIOTest) {
  MemoryDataStore resolver;
  AFF4Flusher<FileBackedObject> file;
  EXPECT_EQ(NewFileBackedObject(&resolver, filename,
                                "truncate", file), STATUS_OK);
  test_Stream(*file);
}


} // namespace aff4
